### ISSUE TITLE
Multi-threaded Scanning

### ISSUE THREAD

**Issue by Cook21** [jbarlow83/OCRmyPDF#509]
**Describe the issue**
While scanning in the task manager I see only one CPU core is Utilized, this makes it even slower than a multi-threaded OCR procedure.

**To Reproduce**

```bash
ocrmypdf --output-type pdf --redo-ocr --jobs 8 --optimize 0 -l eng a.pdf b.pdf
```

**System:**

- OS: Windows 10 version 2004 (Windows Subsystem Linux)
- OCRmyPDF Version: 9.6.1

**Comment by jbarlow83** [comment]
Correct, it is currently single threaded. It is parallelizable but not without some effort.

One issue that pikepdf doesn't allow object marshalling across process boundaries so every core would need to open its own copy of the PDF and do the initial work to determine the PDF is valid, which is still expensive for large files. Only when that overhead becomes negligible would we see benefits from parallelizing, or if object marshalling can be implemented reasonably.

``--redo-ocr`` makes "scanning" more expensive since it requires calculating the position of all text on all pages. This is not normally done.

**Comment by Cook21** [comment]
In my case, Scanning a 88MB PDF file took 1:24:59, by contrast, OCR it only took 17:53 on my 4-core i7-2600K CPU.  I think in some specific scenarios, it is still meaningful to give user a choice on whether or not enable the parallelized scanning even if there is an overhead.

**Comment by jbarlow83** [comment]
I agree it would be nice to have, point is more that some nontrivial development effort is involved.

**Comment by jbarlow83** [comment]
Will be added for next major release.

**Comment by hieu-lam** [comment]
Hi, I just want to say that I had the same issue when running the tool on large PDF files (over 20MB - some could be over 100MB). I used the `--skip-text` flag since some of the pages contain already text. It just simply takes too long time for the initial scanning process to finish. (for some files it takes up to 2+ hours).

You mentioned 

> "Every core should do the initial work to determine the PDF is valid"

What does it mean exactly? We have actually an in-house pipeline process to determine corrupted pdf files, encrypted pdf files, scanned pdf files, scanned pdf files with texts before putting them to ocrmypdf... so we would like this initial scanning process takes less time. :-)

**Comment by jbarlow83** [comment]
@hieu-lam Size of the PDF in MB won't matter; high page count and object count will. 2+ hours sounds excessive. Would be interested to see the files.

There are two things that happen in the "Scan" phase. We use pikepdf to open the PDF and check that it is in fact a PDF and has a valid structure. Most of the heavy lifting is done with libqpdf here and this doesn't normally take long on the file sizes you are referring to. (You can test how long `pikepdf.open()` takes to confirm.) I would not support an effort to bypass this because the inevitable hours would cause all kinds of unreasonable breakage in other places.

The second phase is a two-pass single threaded algorithm to gather information on what is in the PDFs. We need to know the effective DPI and colorspace of all images. The more objects are in the PDF (line art, text, images all count as an object), the longer this takes. I am doing working on both reducing the number of passes and parallelizing it over pages. I suppose some of these checks could be overridden when desired parameters are known.

There was a long term decision to prioritize correctness over speed.

**Comment by hieu-lam** [comment]
@jbarlow83  Thank you for your quick response :-) 

I think it's the second process that takes us a lot of time. Our documents vary between 300 - 1000 pages, some up to 4000 pages. (Yes, I also wonder how can a pdf be this many pages). Initially, we do not know what is in the document, so we run a filter to get those with texts + images, and then use ocrmypdf with flag `--skip-text` with the hope that it will save us some time by only recognizing images. So, back to your explanation, I can see why this takes too long time. 

We can, of course, ignore those big files with texts - but that's not we want. We would like to recognize any scanned images if there is, even if there's only a few out of hundreds of pages. 

Is there any way to solve this? :-)

**Comment by jbarlow83** [comment]
`--skip-text` does save time by only doing OCR on pages that do not have any text, but does not save "scanning" time. There are ways I can improve this, first by parallelizing, which is showing  improvements in my development branches.

If you have some documents with text you really don't want to bypass these checks. If you OCR a page that already has text, you'll duplicate the text on the page, with the OCR version being slightly inferior.

I can consult if that is of interest. I've consulted for several companies for OCRmyPDF, and this has allowed me to develop the software to the maturity it now has. If we sign an agreement then you would then be able to share files with me, and I could identify specific issues and make sure it runs as well as your possible in your environment. I've had some companies like with very large files and some with single page files where accuracy is paramount, and I've done some special customizations and proprietary changes for companies that need it. Your use case sounds sophisticated, complex and interesting.

**Comment by rbrito** [comment]
> `--redo-ocr` makes "scanning" more expensive since it requires calculating the position of all text on all pages. This is not normally done.

For some reason, I don't recall "scanning" taking as much time (in the moderately recent past--perhaps version 8.x?) as it takes currently. Is it possible to do away with at least part of these steps?

I used to use ocrmypdf to optimize scanned books that I download from link.springer.com (essentially, compress multiple BW pages with jbig2).

Just as an extra data point: I noticed that this slowdown occurred more or less when `--verbose 2` started giving more verbose output from `pdfminer`, but this may be a misleading coincidence...

In other words, is there anything possible to make this scan part be abbreviated?

**Comment by rbrito** [comment]
Just for the record, I cloned the repository and started (git-)bisecting it and it seems that with version v8.0.0, the initial (scan) phase is considerably faster than with v9.x with at least the 1st PDF that I am trying (I had too little time to try it already)...

I guess that I should not hijack this issue and create another one talking about this, right?

P.S.: As a side comment, the fact that v8 used to call qpdf before using pdfminer makes me not worry about receiving a dreaded message from pdfminer telling me that some predictors are not supported (and, of course, having to run manually the same qpdf that ocrmypdf would run). Perhaps OCRmyPDF should detect if pdfminer would throw an exception about unsupported predictors and, then, work around this limitation by running qpdf itself? Should this be filed as another feature request/issue?

**Comment by jbarlow83** [comment]
Fixed in v10
